### PR TITLE
[Frontend] Add AutoGraph support for Python for loops

### PR DESCRIPTION
**Pull request by dime10**
Introduces support for Python `for .. in ...:` statement capture as part of the compiled program.
Similar to PR #235, AutoGraph is used to convert such statements into the equivalent Catalyst version before tracing occurs. Specifically, the following constructs are supported via AutoGraph:
- `for elem in iterable:` - These get converted into a `for_loop(0, len(iterable), 1)` with `elem = iterable[i]` automatically assigned using the iteration index, assuming `iterable` is convertible to a JAX array.  If this is not the case, the loop is executed as is in Python.
- `for i in range(start, stop, step):` These are converted directly into their equivalent `for_loop(start, stop, step)`. Contrary to the default Python `range`, when AutoGraph is enabled `range` can also accept dynamic tracers as `start`, `stop`, `step` values. If _any_ exception is raised during the tracing of the `for_loop` body, Catalyst will fall back to Python with a warning.
- `for i, elem in enemurate(iterable):` - These get converted into `for_loop(0, len(iterable), 1)` with the iteration index assigned to the variable chosen by the user (in this case `i`), and `elem = iterable[i]`. This also assumes that `iterable` is convertible to an array, and that the loop body traces without exception, otherwise the loop is executed in Python.

Note that a warning is raised when when a Python fallback is triggered due to a tracing exception. Python fallbacks caused by the iterable not being convertible to array are silent.

[sc-41287]

### CODE REVIEW

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [inline frontend/catalyst/ag_primitives.py:160]: :+1:

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [inline frontend/catalyst/ag_primitives.py:166]: What do you mean by this @dime10?

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [inline frontend/catalyst/ag_primitives.py:330]: is there ever a situation for a user to use this directly?

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [inline frontend/catalyst/ag_primitives.py:362]: It looks like these class properties/methods not tested

File: frontend/catalyst/pennylane_extensions.py
Comment (josh146) [inline frontend/catalyst/pennylane_extensions.py:1394]:
```suggestion
        Catalyst can automatically convert Python for loop statements for you. Requires setting
```

to make the documentation a bit more 'time invariant' :)

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:67]: cool test 😎

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:509]: These are really good integration tests @dime10 👍 

It might make sense though to also add more unit-test style tests. That is, tests that, for a given circuit, compares the autograph output to the expected output.

I'm not sure exactly how this could be done, but perhaps the `autograph_source` function could be used?

The main reason being that integration tests, since they're including circuit simulation, might not always represent accurate conversion (e.g., simple circuits could return the same two values for different structures, potentially because of periodicity, or because of bugs/gates not affecting the state). This is something that has happened a lot in the standard PL test suite over the years.

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:604]: Is this one falling back to python?

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:570]: perhaps good to include a `pytest.warns` here as well

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [inline frontend/test/pytest/test_autograph.py:677]: nice!

@dime10 a lot of your comments here (and in previous files) can be a good starting point for the doc page.

**Review by josh146 — APPROVED** [review]
Thanks @dime10, great work.

I'm approving a bit early because I don't want to be a bottleneck here with my upcoming travel, but I only reviewed the tests and the docs, so the implementation should still be reviewed by someone else :)

Minor docs comment, but the `qjit` docstring should also be updated to specify `for` is now supported:

```rest
autograph (bool): Experimental support for automatically converting Python control
    flow statements to Catalyst-compatible control flow. Currently only supports Python
    ``if``, ``elif``, and ``else`` statements. Note that this feature requires an
    available TensorFlow installation.
```

**Comment by dime10** [comment]
Thanks for the review @josh146 :) Good point on the documentation, I'll have to polish it up a bit before merging!

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:166]: So what I mean is that there are two reasons a fallback to Python could have been triggered:
- the iterable provided by the user is not convertible to an array -> this will fallback to a Python loop silently (without a warning), since there isn't a simple fix to make this loop traceable 
- an exception is raised during the tracing of the loop body after conversion -> this will raise a warning to allow users to correct mistakes and allow the conversion to succeed, for example because they forgot to use a list instead of an array

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:330]: I don't think so? A catalyst for loop doesn't use a range object, a Python loop without autograph would still fail if they used a range with dynamic bounds (whether they use builtin `range` or `CRange`), and a Python loop with autograph already converts the builtin `range` to our `CRange`.

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:362]: That's true. I was thinking of marking this section `nocover`, because it would require testing the entire interface of the builtin `range` function, and all these lines do is forward the API to the builtin `range`.

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:604]: No with this one the conversion succeeds :)

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:570]: Correct 👍

Comment (dime10) [reply frontend/test/pytest/test_autograph.py:570]: I made sure no warnings are raised in this test suite (that we can do something about).

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [reply frontend/catalyst/ag_primitives.py:330]: Yep, makes sense!

File: frontend/catalyst/ag_primitives.py
Comment (josh146) [reply frontend/catalyst/ag_primitives.py:166]: Thanks! Could be worth updating the comment with your github comment, makes it easier to understand :)

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:509]: Hmm yes, we do have unit tests checking the autograph source code transformation over in the [lit tests](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/test/lit/test_autograph.py). I will add something there for the for loop as well, although the testing is less useful than I would like because autograph will just generate `__ag.if_stmt`s or `__ag.for_stmt`s, rather than generating the Catalyst `cond` or `for_loop` directly.

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:828]: Sure 🙂

File: frontend/catalyst/ag_primitives.py
Comment (rmoyard) [inline frontend/catalyst/ag_primitives.py:261]: What is `# nosec B110`?

File: frontend/catalyst/ag_primitives.py
Comment (rmoyard) [reply frontend/catalyst/ag_primitives.py:362]: Would be possible to create this class by inheriting from range? It would remove the testing issue and make things cleaner and more robust to possible python changes.

**Review by rmoyard — COMMENTED** [review]
Great PR :+1: I am just wondering if CRange could inherit from range like you did with enumerate. This would remove issues with testing. Happy to approve after that

File: .github/workflows/check-catalyst.yaml
Comment (dime10) [reply .github/workflows/check-catalyst.yaml:372]: I've added it to `requirements.txt` so I think they should still run

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:134]: 😅 Yeah happy to adjust. The key here is the "non-array" part, a non-array iterable, rather than a non-iterable.

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:261]: I encountered this warning on codefactor, it's not pylint though it's this security checking tool called bandit. The error is complaining about `except: pass`, but I wanted to ensure that this function succeeds no mattter what. The reason being that we call it during an except clause, which we don't want to fail, and we actually don't care care too much if we can't retrieve the original source code information.

Having said that, I can think of a couple of changes we _could_ make, if you prefer:
- catch a specific error: not as safe as catching all errors and continuing without the source code info
- extend the try-except to the entire function body: should be slightly safer even
- move the try-except to where this function is called, rather than inside of it
The last two don't really affect this specific linting message, but they might still be useful anyways.

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:362]: Unfortunately no, it is marked final.

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:509]: So I've ended up adding a bunch of tests which are more unit-testy and which actually uncovered some bugs: https://github.com/PennyLaneAI/catalyst/pull/258/commits/dfd98a7eb296cd6517cadc7ef14a7ab738a2a438
In the process I also improved the error checking:
https://github.com/PennyLaneAI/catalyst/pull/258/commits/8d2a40ce606d7d1206ddecd67924783da047354a

Let me know if you think this is sufficient :)

File: frontend/test/pytest/test_autograph.py
Comment (dime10) [reply frontend/test/pytest/test_autograph.py:828]: Added here: https://github.com/PennyLaneAI/catalyst/pull/258/commits/c86b95785e025ee9816971dff731bea886a0648c

File: frontend/catalyst/pennylane_extensions.py
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:1394]: https://github.com/PennyLaneAI/catalyst/pull/258/commits/063e547880be0260727b13b49cbbb8a5e30a3c38

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:362]: For now I've marked the `range` interface itself as `nocover` (for the above reasons), but I did add a unit test that checks the fallback of `CRange` to `range` is working correctly: https://github.com/PennyLaneAI/catalyst/pull/258/commits/5341c3abc09d5f87a539ee99783487e4596786de

Let me know if this is acceptable :)

File: frontend/catalyst/ag_primitives.py
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:166]: https://github.com/PennyLaneAI/catalyst/pull/258/commits/515dbcfc2704b676dc275d78238494a61ddc74e3

**Comment by dime10** [comment]
I've added an additional user facing feature here: https://github.com/PennyLaneAI/catalyst/pull/258/commits/8219ffa7ca7414cfd693fd7efededc5e51fe77d9

It provides two flags that the user can set around the conversion strictness (I used them during debugging and in the tests):
- `catalyst.autograph_ignore_fallbacks`: Silences warnings resulting from Python fallbacks.
  This is useful if the user doesn't want to see any warnings, or in situations where the warning cannot be fixed.
- `catalyst.autograph_strict_conversion`: Turns Python fallback warnings into errors and produces the full
  traceback of the error that caused the fallback.
  This can be useful when debugging why a fallback happened.

File: frontend/test/pytest/test_autograph.py
Comment (josh146) [reply frontend/test/pytest/test_autograph.py:522]: Nice, I love to hear that 💪

**Comment by josh146** [comment]
Thanks @dime10! I think these are good options to include.

Non-blocking for this PR, but for end-users I worry about having these new options be module variables as opposed to an option you pass to `@qjit`.

In Strawberry Fields, we used to have `strawberryfields.hbar` be a module level variable that users could set, and we ran into a tonne of problems and edge cases due to this global state/the fact that functions weren't pure.

It mainly showed up in several places:

- Parallelization. E.g., users using threads, or dask, or even pytest running tests in parallel, would cause race conditions.

- developers would sometimes alter these values internally, and forget to change them back. Or there would be an exception before it could change back (and the developer didn't use `try`-`finally`).

So before usage gets baked in, might be good to turn these into arguments.

**Review by rmoyard — APPROVED** [review]
It looks good to me :100: I suggest a autograph configuration kwarg for qjit but not blocking

**Comment by dime10** [comment]
I was originally thinking of similar configurations in JAX which are also global, like `jax.config.update("jax_debug_nans", True)`. However, you raise some good points so I'm thinking we could add something like this:
```py
@qjit(autograph=True, ag_mode=...)
```
Where `ag_mode` can be one of the following:
  - "strict": Always raise an error with stacktrace when a conversion fails. Enable this
      to debug why a conversion may have failed, or when successful conversion is
      critical.
  - "permissive": The default & safe option. Catalyst will never abort compilation because
      an AutoGraph conversion failed, but some errors are re-raised as warnings.
  - "silent": Catalyst will not raise any warnings or errors and will fall back to Python
      silently whenever AutoGraph conversion fails.

However, there is a problem because there is nothing actually linking the qjit object to the eventual invocation of the autograph transformed for loop. I think this is true in general for all our primitives, so there is no obvious to pass compilation options into any of them.

**Comment by josh146** [comment]
> - "strict": Always raise an error with the full traceback when a conversion fails. Enable this
>   to debug why a conversion may have failed, or when successful conversion is critical.
> - "permissive": The default & safe option. Catalyst will never abort compilation because
>   an AutoGraph conversion failed, but some errors are re-raised as warnings.
> - "silent": Catalyst will not raise any warnings or errors and will fall back to Python
>   silently whenever AutoGraph conversion fails.

I really like this 🙌 I'll add it to Q4 as a potential story.

**Pull request merged**
